### PR TITLE
Sinope TH1400ZB converter enhancement

### PR DESCRIPTION
**Pull request by faucherc**
I have add and adjust all the setting available for the Sinope TH1400ZB. The final result fits the Sinope app and/or documentation.

### CODE REVIEW

File: converters/fromZigbee.js [hunk 1]
```diff
@@ -3480,14 +3484,84 @@ const converters = {
         cluster: 'hvacThermostat',
         type: ['readResponse'],
         convert: (model, msg, publish, options, meta) => {
-            const lookup = {0: 'unoccupied', 1: 'occupied'};
-            const lookup1 = {0: 'on_demand', 1: 'sensing'};
             const result = {};
+            const cycleOutputLookup = {15: '15 sec', 300: '5 min', 600: '10 min',
```
Comment (Koenkk) [inline converters/fromZigbee.js:3488]: Can you make this snake case, e.g. `15 sec` -> `15_sec`

File: converters/fromZigbee.js [hunk 2]
```diff
@@ -3480,14 +3484,84 @@ const converters = {
...
+        convert: (model, msg, publish, options, meta) => {
+            const result = {};
+            if (msg.data.hasOwnProperty('outdoorTempToDisplay')) {
+                result.outdoor_temp_to_display = msg.data['outdoorTempToDisplay'];
+            }
+            if (msg.data.hasOwnProperty('outdoorTempToDisplayTimeout')) {
+                result.enable_outdoor_temperature = msg.data['outdoorTempToDisplayTimeout'] == 10800 ? 'on' : 'off';
+            }
+            if (msg.data.hasOwnProperty('currentTimeToDisplay')) {
+                result.currentTimeToDisplay = msg.data['currentTimeToDisplay'];
```
Comment (Koenkk) [inline converters/fromZigbee.js:3519]: `currentTimeToDisplay` -> `current_time_to_display`

File: converters/fromZigbee.js [hunk 3]
```diff
@@ -3480,14 +3484,84 @@ const converters = {
...
+            return result;
+        },
+    },
+    sinope_TH1400ZB_specific: {
+        cluster: 'manuSpecificSinope',
+        type: ['readResponse'],
+        convert: (model, msg, publish, options, meta) => {
+            const result = {};
+            if (msg.data.hasOwnProperty('outdoorTempToDisplay')) {
+                result.outdoor_temp_to_display = msg.data['outdoorTempToDisplay'];
```
Comment (Koenkk) [inline converters/fromZigbee.js:3513]: `outdoor_temp_to_display` -> `outdoor_temperature_to_display`

File: converters/fromZigbee.js [hunk 4]
```diff
@@ -3480,14 +3484,84 @@ const converters = {
...
+                if (result.floor_max_heat_setpoint === -327.68) {
+                    result.floor_max_heat_setpoint = 'off';
+                }
+            }
+            if (msg.data.hasOwnProperty('temperatureSensor')) {
+                const lookup = {0: '10k', 1: '12k'};
+                result.floor_temperature_sensor = lookup[msg.data['temperatureSensor']];
+            }
+            if (msg.data.hasOwnProperty('floorLimitStatus')) {
+                result.floorLimitStatus = msg.data['floorLimitStatus'];
```
Comment (Koenkk) [inline converters/fromZigbee.js:3548]: `floorLimitStatus` -> `floor_limit_status`

File: converters/fromZigbee.js [hunk 5]
```diff
@@ -3480,14 +3484,84 @@ const converters = {
...
+                result.connected_load = msg.data['connectedLoad'];
+            }
+            if (msg.data.hasOwnProperty('auxConnectedLoad')) {
+                result.aux_connected_load = msg.data['auxConnectedLoad'];
+                if (result.aux_connected_load == 65535) {
+                    result.aux_connected_load = 'disabled';
+                }
+            }
+            if (msg.data.hasOwnProperty('pumpProtection')) {
+                result.pump_protection = msg.data['pumpProtection'] == 1 ? 'on' : 'off';
```
Comment (Koenkk) [inline converters/fromZigbee.js:3564]: `on`/`off` -> `ON`/`OFF`

File: converters/fromZigbee.js (hunk 1 above)
Comment (faucherc) [reply converters/fromZigbee.js]: Done

File: converters/fromZigbee.js (hunk 2 above)
Comment (faucherc) [reply converters/fromZigbee.js]: Attribute removed. I forgot to remove it.

File: converters/fromZigbee.js (hunk 3 above)
Comment (faucherc) [reply converters/fromZigbee.js]: Attribute removed. I forgot to remove it.

File: converters/fromZigbee.js (hunk 4 above)
Comment (faucherc) [reply converters/fromZigbee.js]: Attribute removed. I forgot to remove it.

File: converters/fromZigbee.js (hunk 5 above)
Comment (faucherc) [reply converters/fromZigbee.js]: Done

**Comment by MattL0** [comment]
@faucherc thanks for this, i might check to change some thingas in 1124zb from this. 

Can i ask you something ? what is main_cycle_output ? Maybe it is an option only for floor thermostats... but i am able to write a double to that only if it starts swith 15. So 15.15 would work. I am also ble to read this attribute. 

Would it have any use for a heat only thermostat?

**Comment by faucherc** [comment]
@MattL0 I also have some 1123zb. I will modify the converter for it like I have done with the 1400zb.  The 1123 and 1124 are two little brothers.  So I guest that the modification I will made for the 1123 can also be applied to the 1124.  If you are interested, we can work together to test both of them.

In the Sinope App, in the advanced configuration for the 1123, I have "Main Cycle Length" with possible values "short cycle" and "long cycle".  If you are able to read it, I'm suspecthing that Sinope used the "main_cycle_output" for this param.  

It seems that we can modify this params directly on the thermostat. (https://support.sinopetech.com/en/1.1.2.4/) You can modify it on the thermostat and seen if the value has changed when you try to read it again.   If I'm right, modifying and reading the value for both available value, we will have the values to use it in the converter.

**Comment by MattL0** [comment]
i sent a pr to this https://github.com/faucherc/zigbee-herdsman-converters/pull/1

**Comment by MattL0** [comment]
may i suggest to send all really spécific exposes to option ? 

like this 

        zigbeeModel: ['TH1124ZB'],
        model: 'TH1124ZB',
        vendor: 'Sinopé',
        description: 'Zigbee line volt thermostat',
        meta: {thermostat: {dontMapPIHeatingDemand: true}},
        fromZigbee: [fz.legacy.sinope_thermostat_att_report, fz.legacy.hvac_user_interface, fz.electrical_measurement, fz.metering,
            fz.ignore_temperature_report, fz.legacy.sinope_thermostat_state, fz.sinope_thermostat, fz.ignore_genOta],

        toZigbee: [tz.thermostat_local_temperature, tz.thermostat_occupied_heating_setpoint, tz.thermostat_unoccupied_heating_setpoint,
            tz.thermostat_temperature_display_mode, tz.thermostat_keypad_lockout, tz.thermostat_system_mode, tz.thermostat_running_state,
            tz.sinope_thermostat_backlight_autodim_param, tz.sinope_thermostat_time, tz.sinope_thermostat_enable_outdoor_temperature,
            tz.sinope_thermostat_outdoor_temperature, tz.sinope_time_format, tz.thermostat_min_heat_setpoint_limit,
            tz.thermostat_max_heat_setpoint_limit, tz.abs_thermostat_min_heat_setpoint_limit, tz.abs_thermostat_max_heat_setpoint_limit,
            tz.sinope_thermostat_main_cycle_output, tz.sinope_thermostat_occupancy],
        options: [exposes.numeric('min_heat_setpoint_limit', ea.ALL)
                .withUnit('°C')
                .withValueMin(5)
                .withValueMax(30)
                .withValueStep(0.5)
                .withDescription('Occupied Heating setpoint Lower limit'),
            exposes.numeric('max_heat_setpoint_limit', ea.ALL)
                .withUnit('°C')
                .withValueMin(5)
                .withValueMax(30)
                .withValueStep(0.5)
                .withDescription('Occupied Heating setpoint Higher limit'),
            exposes.numeric('abs_min_heat_setpoint_limit', ea.ALL)
                .withUnit('°C')
                .withValueMin(5)
                .withValueMax(30)
                .withValueStep(0.5)
                .withDescription('Unoccupied Heating setpoint Lower limit'),
            exposes.numeric('abs_max_heat_setpoint_limit', ea.ALL)
                .withUnit('°C')
                .withValueMin(5)
                .withValueMax(30)
                .withValueStep(0.5)
                .withDescription('Unoccupied Heating setpoint Higher limit')],
        exposes: [
            exposes.climate()
                .withLocalTemperature()
                .withSetpoint('occupied_heating_setpoint', 5, 30, 0.5)
                .withSetpoint('unoccupied_heating_setpoint', 5, 30, 0.5)
                .withSystemMode(['off', 'heat'])
                .withRunningState(['idle', 'heat'])
                .withPiHeatingDemand(),
            exposes.enum('thermostat_occupancy', ea.set, ['unoccupied', 'occupied'])
                .withDescription('Occupancy state of the thermostat'),
            exposes.enum('backlight_auto_dim', ea.ALL, ['on demand', 'sensing'])
                .withDescription('Control backlight dimming behavior'),
            exposes.enum('keypad_lockout', ea.ALL, ['unlock', 'lock1'])
                .withDescription('Enables or disables the device’s buttons'),
            e.power(), e.current(), e.voltage(), e.energy(),
    ],

**Comment by faucherc** [comment]
@Koenkk I have update my branch with Master.  Is there any adjustment or modification I can made to make your review more easy and finally get your approval?

File: converters/toZigbee.js [hunk 6]
```diff
@@ -4252,6 +4256,18 @@ const converters = {
             await entity.read('hvacThermostat', [0x4000], manufacturerOptions.eurotronic);
         },
     },
+    sinope_thermostat_report_local_temperature: {
+        key: ['report_local_temperature'],
+        convertSet: async (entity, key, value, meta) => {
+            const reportLocalTemperatureValues = {0: 'automatic', 1: '', 2: 'outside temp'};
```
Comment (Koenkk) [inline converters/toZigbee.js:4262]: `outside temp` -> `outside_temperature`

Comment (Koenkk) [reply converters/toZigbee.js:4262]: I also noticed there is no expose for this.

**Comment by Koenkk** [comment]
I found one more issue, after this is fixed it can be merged.

File: converters/toZigbee.js (hunk 6 above)
Comment (faucherc) [reply converters/toZigbee.js]: This was not used and was not necessary. I have removed it. I have check all the added methods to be sure that's the only one.

**Comment by faucherc** [comment]
@Koenkk I have update the branch from master and remove the obselete method.

**Comment by Koenkk** [comment]
Thanks!

**Pull request merged**